### PR TITLE
Marching cubes

### LINKED ISSUE #21: Draw 3D object by its equation F(x,y,z) = 0
**Is your feature request related to a problem? Please describe.**
We want to draw 3D objects by its equation `F(x,y,z) = 0`. This feature is needed to draw objects inserted into considered region.

**Describe the solution you'd like**
I suggest to use well-known `Marching Cubes` algorithm.
Useful refs:
https://en.wikipedia.org/wiki/Marching_cubes
https://graphics.stanford.edu/~mdfisher/MarchingCubes.html
https://github.com/nsf/mc
https://github.com/dominikwodniok/dualmc

### PR DESCRIPTION
**Pull request by Crosby98**
# What this PR does?
Implements a Draw of 3D object by its equation

## Issue reference
Fixes #21

## Explanation
We want to draw 3D objects by its equation `F(x,y,z) = 0`. This feature is needed to draw objects inserted into considered region.

# How to get?
`git clone --recursive -b marching-cubes git@github.com:aartiukh/sph.git`

### CODE REVIEW

File: algorithms/src/MarchingCubes.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+#include "MarchingCubes.h"
+#include "MarchingCubesConfig.h"
+
+#include <vector>
+
+namespace SPHAlgorithms
+{
+
+std::vector<Point3F> MarchingCubes::getFunctionMesh(std::function<float(float, float, float)> f)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
Point3FVector  MarchingCubes::getFunctionMesh(std::function<float(float, float, float)> f)
```

Comment (aartiukh) [reply algorithms/src/MarchingCubes.cpp]: Here and check below please

File: algorithms/src/MarchingCubes.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        CubeValue[iVertex] =
+            f(fX + VertexOffset[iVertex][0], fY + VertexOffset[iVertex][1], fZ + VertexOffset[iVertex][2]);
+    }
+
+    // Flag that indicates that vertice is inside or outsice of the surface
+    int iFlagIndex = 0;
+
+    // Find which vertices are inside of the surface and which are outside
+    iFlagIndex = determineFlag(iFlagIndex, CubeValue);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Combine to one line with declaration

File: algorithms/src/MarchingCubes.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    int iEdgeFlags = CubeEdgeFlags[iFlagIndex];
+
+    // If the cube is entirely inside or outside of the surface, then there will be no intersections
+    if (iEdgeFlags == 0)
+    {
+        return resultTrianglesMesh;
+    }
+
+    // Points of intersection of the surface with each edge
+    std::vector<Point3F> EdgeVertex(12);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Magic constant `12`

File: algorithms/src/MarchingCubes.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    // Find points of intersection on each edge
+    findPointIntersection(EdgeVertex, iEdgeFlags, CubeValue, fX, fY, fZ);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Should return `EdgeVertex`

File: algorithms/src/MarchingCubes.cpp [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    // Fill the triangles that were found.  There can be up to five per cube
+    fillFoundTriangles(resultTrianglesMesh, EdgeVertex, iFlagIndex);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Should return `resultTrianglesMesh`

File: algorithms/src/MarchingCubes.cpp [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+    return resultTrianglesMesh;
+}
+
+void MarchingCubes::fillFoundTriangles(std::vector<Point3F>&       resultEdgeVertex,
+                                       const std::vector<Point3F>& EdgeVertex,
+                                       const int                   iFlagIndex)
+{
+    for (int iTriangle = 0; iTriangle < 5; iTriangle++)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `5` is a magic number

File: algorithms/src/MarchingCubes.cpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        if (TriangleConnectionTable[iFlagIndex][3 * iTriangle] < 0)
+        {
+            break;
+        }
+
+        for (int iCorner = 0; iCorner < 3; iCorner++)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `3` is a magic number

File: algorithms/src/MarchingCubes.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    /**
+     * @brief Generates triangles mesh from function
+     * @param f    The function that represents the domain equation
+     */
+    static std::vector<Point3F> getFunctionMesh(std::function<float(float, float, float)> f);
+
+private:
+    static std::vector<Point3F> MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ);
+
+    static void generateObjFile(const std::vector<Point3F>& vertices, const std::string& fileName);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.h]: Remove

File: algorithms/test/src/MarchingCubesTestSuite.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    generateObjFile(mesh, "pawn.obj");
+}
+
+} // namespace TestEnvironment
+} // namespace SPHAlgorithms
+
+using namespace SPHAlgorithms::TestEnvironment;
+
+TEST(MarchingCubesTestSuite, generatePawnMesh)
```
Comment (aartiukh) [inline algorithms/test/src/MarchingCubesTestSuite.cpp:107]: Add test for bishop

File: algorithms/src/MarchingCubes.cpp [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+    return -fValue1 / fDelta;
+}
+
+std::vector<Point3F> MarchingCubes::MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ)
+{
+    // vector with all triangles found
+    std::vector<Point3F> resultTrianglesMesh;
+
+    float CubeValue[8];
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `8` is a magic constant

File: algorithms/src/MarchingCubes.cpp [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+#include <vector>
+
+namespace SPHAlgorithms
+{
+
+std::vector<Point3F> MarchingCubes::getFunctionMesh(std::function<float(float, float, float)> f)
+{
+    std::vector<Point3F> resultMesh;
+
+    for (float iX = X_MIN; iX < X_MAX; iX += GRID_CUBE_SIZE)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Check if we reach `X_MAX` with this `for` loop

File: algorithms/src/MarchingCubes.cpp [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    {
+        for (float iY = Y_MIN; iY < Y_MAX; iY += GRID_CUBE_SIZE)
+        {
+            for (float iZ = Z_MIN; iZ < Z_MAX; iZ += GRID_CUBE_SIZE)
+            {
+                // iX, iY, iZ are coordinates of the current vertex in grid_cube
+                auto vertices = MarchCube(f, iX, iY, iZ);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
                const auto vertices = MarchCube(f, x, y, z);
```

File: algorithms/src/MarchingCubes.cpp [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+#include "MarchingCubesConfig.h"
+
+#include <vector>
+
+namespace SPHAlgorithms
+{
+
+std::vector<Point3F> MarchingCubes::getFunctionMesh(std::function<float(float, float, float)> f)
+{
+    std::vector<Point3F> resultMesh;
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
    std::vector<Point3F> mesh;
```

File: algorithms/src/MarchingCubes.cpp (hunk 11 above)
Comment (aartiukh) [reply algorithms/src/MarchingCubes.cpp]:
```suggestion
    for (float x = X_MIN; x < X_MAX; x += GRID_CUBE_SIZE)
```

File: algorithms/src/MarchingCubes.cpp [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+    for (float iX = X_MIN; iX < X_MAX; iX += GRID_CUBE_SIZE)
+    {
+        for (float iY = Y_MIN; iY < Y_MAX; iY += GRID_CUBE_SIZE)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
        for (float y = Y_MIN; y < Y_MAX; y += GRID_CUBE_SIZE)
```

File: algorithms/src/MarchingCubes.cpp [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        {
+            for (float iZ = Z_MIN; iZ < Z_MAX; iZ += GRID_CUBE_SIZE)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
            for (float z = Z_MIN; z < Z_MAX; z += GRID_CUBE_SIZE)
```

File: algorithms/src/MarchingCubes.cpp [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+                // iX, iY, iZ are coordinates of the current vertex in grid_cube
+                auto vertices = MarchCube(f, iX, iY, iZ);
+                std::copy(vertices.begin(), vertices.end(), std::back_inserter(resultMesh));
+            }
+        }
+    }
+    return resultMesh;
+}
+
+static float adapt(float fValue1, float fValue2)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
static float adapt(float a, float b)
```

File: algorithms/src/MarchingCubes.cpp [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+{
+    const float fDelta = fValue2 - fValue1;
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
    const float delta = a - b;
```

File: algorithms/src/MarchingCubes.cpp [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+    if (std::abs(fDelta) < 0.0001f)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: Make a constant for `0.0001f`

File: algorithms/src/MarchingCubes.cpp [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        return 0.5f;
+    }
+
+    return -fValue1 / fDelta;
+}
+
+std::vector<Point3F> MarchingCubes::MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ)
+{
+    // vector with all triangles found
+    std::vector<Point3F> resultTrianglesMesh;
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]:
```suggestion
    std::vector<Point3F> trianglesMesh;
```

File: algorithms/src/MarchingCubes.cpp [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+void MarchingCubes::findPointIntersection(std::vector<Point3F>& EdgeVertex,
+                                          const int             iEdgeFlags,
+                                          const float           CubeValue[],
+                                          const float           fX,
+                                          const float           fY,
+                                          const float           fZ)
+{
+    for (int iEdge = 0; iEdge < 12; iEdge++)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `12` is a magic number

File: algorithms/src/MarchingCubes.h [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+ * @author Ihor Lytvyn (litvinigor98@gmail.com)
+ * @date Created May 26, 2019
+ **/
+
+#ifndef MARCHING_CUBES_H_43C34465A6ED4DB9B9F2F4C3937BF5DC
+#define MARCHING_CUBES_H_43C34465A6ED4DB9B9F2F4C3937BF5DC
+
+#include "Point.h"
+#include <functional>
+#include <vector>
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.h]: Remove

File: algorithms/src/MarchingCubes.h [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+public:
+    /**
+     * @brief Generates triangles mesh from function
+     * @param f    The function that represents the domain equation
+     */
+    static std::vector<Point3F> getFunctionMesh(std::function<float(float, float, float)> f);
+
+private:
+    static std::vector<Point3F> MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.h]:
```suggestion
    static std::vector<Point3F> MarchingCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ);
```

File: algorithms/src/MarchingCubes.h [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+                                   const int                   iFlagIndex);
+
+    static void findPointIntersection(std::vector<Point3F>& EdgeVertex,
+                                      const int             iEdgeFlags,
+                                      const float           CubeValue[],
+                                      const float           fX,
+                                      const float           fY,
+                                      const float           fZ);
+
+    static int determineFlag(int flag, const float CubeValue[]);
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.h]:
```suggestion
    static int determineFlag(const float CubeValue[]);
```

File: algorithms/src/MarchingCubes.cpp [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+
+std::vector<Point3F> MarchingCubes::MarchCube(std::function<float(float, float, float)> f, float fX, float fY, float fZ)
+{
+    // vector with all triangles found
+    std::vector<Point3F> resultTrianglesMesh;
+
+    float CubeValue[8];
+
+    // Evaluate value of the cube vertex at each point
+    for (int iVertex = 0; iVertex < 8; iVertex++)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `8` is a magic constant

**Review by aartiukh — CHANGES_REQUESTED** [review]

File: algorithms/src/MarchingCubes.cpp (hunk 5 above)
Comment (Crosby98) [reply algorithms/src/MarchingCubes.cpp]: i dont think that it should return something

File: algorithms/src/MarchingCubes.cpp [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }
+    return mesh;
+}
+
+static float adapt(float a, float b)
+{
+    const float delta = b - a;
+
+    if (std::abs(delta) < PRESIZION)
```
Comment (aartiukh) [inline algorithms/src/MarchingCubes.cpp]: `PRECISION` :)

File: algorithms/src/MarchingCubes.h (hunk 21 above)
Comment (aartiukh) [reply algorithms/src/MarchingCubes.h]: Not fixed

File: algorithms/test/src/MarchingCubesTestSuite.cpp [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+namespace SPHAlgorithms
+{
+namespace TestEnvironment
+{
+
+void MarchingCubesTestSuite::generatePawnMesh()
+{
+    const std::vector<Point3F> mesh = SPHAlgorithms::MarchingCubes::getFunctionMesh(pawnEquation);
```
Comment (aartiukh) [inline algorithms/test/src/MarchingCubesTestSuite.cpp]: Use typedef

File: algorithms/test/src/MarchingCubesTestSuite.cpp [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+
+    ASSERT_EQ(37128u, mesh.size());
+
+    generateObjFile(mesh, "pawn.obj");
+}
+
+void MarchingCubesTestSuite::generateBishopMesh()
+{
+	const std::vector<Point3F> mesh = SPHAlgorithms::MarchingCubes::getFunctionMesh(bishopEquation);
```
Comment (aartiukh) [inline algorithms/test/src/MarchingCubesTestSuite.cpp]: Use typedef

**Review by aartiukh — CHANGES_REQUESTED** [review]

**Review by aartiukh — DISMISSED** [review]

**Review by aartiukh — APPROVED** [review]

**Pull request merged**